edgelet/edgelet-http/src/unix.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+#![cfg(unix)]
+
+use std::fs;
+use std::os::unix::fs::MetadataExt;
+use std::path::Path;
+
+use nix::sys::stat::{umask, Mode};
+use tokio_core::reactor::Handle;
+use tokio_uds::UnixListener;
+
+use error::Error;
+use util::incoming::Incoming;
+
+pub fn listener<P: AsRef<Path>>(path: P, handle: &Handle) -> Result<Incoming, Error> {
+    let listener = if path.as_ref().exists() {
+        // get the previous file's metadata
+        let metadata = fs::metadata(&path)?;
+        debug!("read metadata {:?} for {}", metadata, path.as_ref().display());
+
+        debug!("unlinking {}...", path.as_ref().display());
+        fs::remove_file(&path)?;
+        debug!("unlinked {}", path.as_ref().display());
+
+        let mode = Mode::from_bits_truncate(metadata.mode());
+        let mut mask = Mode::all();
+        mask.toggle(mode);
+
+        debug!("settings permissions {:#o} for {}...", mode, path.as_ref().display());
+        let prev = umask(mask);
+        defer! {{ umask(prev); }}
```

Comment:
I am guessing this restores the current process's umask when this function returns? Nice.

Reply:
Yes. It's similar to go's `defer`. The macro creates a struct and implements `Drop` and calls the block given to the macro in the `drop`.

Reply:
Very cool!